### ISSUE TITLE
confusing error: The argument type 'dart.core::String' can't be assigned to the parameter type 'dart.core::String'.

### ISSUE THREAD

**Issue by mraleph** [dart-lang/sdk#35715]
Flutter users reporting bizarre errors from the compiler that confusingly appear and disappear (might be related to serialized incremental compiler state?)

```
Compiler message:
lib/main.dart:73:35: Error: The argument type 'dart.core::String' can't be assigned to the parameter type 'dart.core::String'.
Try changing the type of the parameter, or casting the argument to 'dart.core::String'.
            image: new AssetImage('s'),
```

I don't have a concrete reproduction for it - but I asked users that face it to send us their Dart source if possible. 

I saw some reports that users resolved it by changing import scheme from relative to package-scheme imports

Any ideas what can be causing this @peter-ahe-google @kmillikin @jensjoha ? Maybe a bug in class hierarchy? 

Maybe we could add some debugging information to CFE to dump more debugging information for this sort of error?

**Comment by zoechi** [comment]
After #33076 was fixed I have only seen this mentioned on Windows where upper-case wasn't used consistently when importing the same file.
@DanTup encountered this case a few times AFAIK with users.

**Comment by lrhn** [comment]
The import-case issue explains other classes, but not `String` from `dart:core`. You cannot import `dart:core` with a different case. There must be something more fundamentally wrong if you have two versions of `dart:core` in your program.

**Comment by peter-ahe-google** [comment]
We are exploring a few ideas for how this situation could arise.

**Comment by peter-ahe-google** [comment]
We have a reproduction. More details will follow.

**Comment by peter-ahe-google** [comment]
@aam let me know if you need to discuss this in detail. I think it may be a simple fix: Make sure that `CompilerOptions.sdkSummary` is null as well as `CompilerOptions.inputSummaries` and `CompilerOptions.linkedDependencies` are both empty when using `IncrementalCompiler.fromComponent`.

**Comment by aam** [comment]
It seems that we need to ensure that saved last known good state includes platform libraries, which doesn't currently - when we rebuild incremental compiler we reload platform libraries so references from saved state don't match to what is loaded for new incremental compiler.

@jensjoha  @peter-ahe-google can you please take a look at concept https://dart-review.googlesource.com/c/sdk/+/91859 which basically adds list of platform libraries that incremental compiler can stash and use when reject has to rebuild compiler.

**Comment by a-siva** [comment]
is this issue resolved?

**Comment by aam** [comment]
> is this issue resolved?

No, this is work in progress as this tracks proper implementation of vm `reject` request.

There are changes that @jensjoha implementing that will make fasta incremental compiler respect `--omit-platform` flag, then vm incremental compiler will be requesting incremental components with platform libraries, filter them out(based on `dart:` scheme) and add them to incremental component when `reject` request has to be processed.

**Comment by jensjoha** [comment]
Should have been fixed by 2087d6db46c3b3508e2007191ba9f5e88f32a37d.
